crates/guest-rust/rt/src/async_support/future_support.rs

```diff
@@ -61,227 +238,243 @@ impl<T> fmt::Debug for FutureWriter<T> {
+        self.pin_project()
+            .poll_complete(cx)
+            .map(|(result, _writer)| match result {
+                WriteComplete::Written => Ok(()),
+                WriteComplete::Closed(value) | WriteComplete::Cancelled(value) => {
+                    Err(FutureWriteError { value })
                 }
-                Handle::Read => unreachable!(),
-                Handle::Write | Handle::LocalClosed => unsafe {
-                    entry.remove();
-                    (self.vtable.close_writable)(self.handle, 0);
-                },
-                Handle::WriteClosedErr(_) => match entry.remove() {
-                    Handle::WriteClosedErr(None) => unsafe {
-                        (self.vtable.close_writable)(self.handle, 0);
-                    },
-                    Handle::WriteClosedErr(Some(err_ctx)) => unsafe {
-                        (self.vtable.close_writable)(self.handle, err_ctx.handle());
-                    },
-                    _ => unreachable!(),
-                },
-            },
-        });
+            })
     }
 }
 
-/// Represents a read operation which may be canceled prior to completion.
-pub struct CancelableRead<T: 'static> {
-    reader: Option<FutureReader<T>>,
-    future: Pin<Box<dyn Future<Output = Option<Result<T, ErrorContext>>>>>,
+impl<T: 'static> FutureWrite<T> {
+    fn pin_project(self: Pin<&mut Self>) -> Pin<&mut WaitableOperation<FutureWriteOp<T>>> {
+        // SAFETY: we've chosen that when `Self` is pinned that it translates to
+        // always pinning the inner field, so that's codified here.
+        unsafe { Pin::new_unchecked(&mut self.get_unchecked_mut().op) }
+    }
+
+    /// Cancel this write if it hasn't already completed.
+    ///
+    /// This method can be used to cancel a write-in-progress and re-acquire
+    /// the writer and the value being sent. If the write operation has already
+    /// succeeded racily then `None` is returned and the write completed.
+    ///
+    /// Possible return values are:
+    ///
+    /// * `Ok(())` - the pending write completed before cancellation went
+    ///   through meaning that the original message was actually sent.
+    /// * `Err(FutureWriteCancel::Closed(v))` - the pending write did not complete
+    ///   because the other end was closed before receiving the value. The value
+    ///   is provided back here as part of the error.
+    /// * `Err(FutureWriteCancel::Cancelled(v, writer))` - the pending write was
+    ///   cancelled. The value `v` is returned back and the `writer` is returned
+    ///   as well to resume a write in the future if desired.
+    ///
+    /// Note that if this method is called after the write was already cancelled
+    /// then `Ok(())` will be returned.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the operation has already been completed via `Future::poll`,
+    /// or if this method is called twice.
+    pub fn cancel(self: Pin<&mut Self>) -> Result<(), FutureWriteCancel<T>> {
```

Comment:
What's the rationale for using `Result` here rather than adding a `FutureWriteCancel::AlreadyDelivered` variant and having this return `FutureWriteCancel<T>` instead?  It's a little surprising to me that we'd consider `FutureWriteCancel::Cancelled` (or really any of the possible outcomes) to be an error.